### PR TITLE
feat: add new locales in authoring

### PR DESCRIPTION
**Pull request by gabrielfs7**
⚠️ released at https://github.com/oat-sa/tao-core/pull/4139

https://oat-sa.atlassian.net/browse/ADF-1796
https://oat-sa.atlassian.net/browse/HKD-6

# Goal

- Support new locales (16 new locales were introduced): so-SO,ti-ER,am-ET,tr-TR,ug-CN,my-MM,th-TH,ku-IQ,vi-VN,sq-AL,bs-BA,sw-KE,ta-IN,tl-PH,hr-HR,sr-RS
- Allow to restrict translation locales based on feature flag `TAO_ALLOWED_TRANSLATION_LOCALES`

# How to test

- [ ] Set the .env with `TAO_ALLOWED_TRANSLATION_LOCALES=en-US,fr-FR`
- [ ] Check the in the UI language both in Item authoring and Form will be restricted to `en-US` and `fr-FR`

Screenshot 2024-11-15 at 14 34 26

Screenshot 2024-11-15 at 14 37 47

### CODE REVIEW

File: migrations/Version202411141300522230_tao.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+final class Version202411141300522230_tao extends AbstractMigration
+{
+    public function getDescription(): string
+    {
+        return 'Sync models';
+    }
+
+    public function up(Schema $schema): void
+    {
+        OntologyUpdater::syncModels();
```
Comment (shpran) [inline migrations/Version202411141300522230_tao.php:23]: I think it's time to create a ticket for this and do this synchronization in `tao/scripts/taoUpdate` script

File: models/classes/Language/Filter/LanguageAllowedFilter.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+use oat\tao\model\Language\LanguageCollection;
+use oat\tao\model\Lists\Business\Domain\ValueCollection;
+use oat\tao\model\TaoOntology;
+
+class LanguageAllowedFilter
+{
+    public const TAO_ALLOWED_TRANSLATION_LOCALES = 'TAO_ALLOWED_TRANSLATION_LOCALES';
+
+    /** @var tao_models_classes_LanguageService */
+    private $allowedLocales;
```
Comment (shpran) [inline models/classes/Language/Filter/LanguageAllowedFilter.php:37]:
```suggestion
    /** @var string[] */
    private array $allowedLocales;
```

File: models/classes/Language/Filter/LanguageAllowedFilter.php [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    {
+        if (empty($this->allowedLocales)) {
+            return $collection;
+        }
+
+        $indicesToUnset = [];
+
+        /** @var Language $item */
+        foreach ($collection as $key => $item) {
+            if (!in_array($item->getCode(), $this->allowedLocales)) {
```
Comment (shpran) [inline models/classes/Language/Filter/LanguageAllowedFilter.php:63]:
```suggestion
            if (!in_array($item->getCode(), $this->allowedLocales, true)) {
```

File: models/classes/Language/Filter/LanguageAllowedFilter.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    public function filterByValueCollection(ValueCollection $collection): ValueCollection
+    {
+        if (empty($this->allowedLocales)) {
+            return $collection;
+        }
+
+        $allowedUris = [];
+
+        foreach ($this->allowedLocales as $locale) {
+            $allowedUris[] = sprintf('%s%s', TaoOntology::LANGUAGE_PREFIX, $locale);
```
Comment (shpran) [inline models/classes/Language/Filter/LanguageAllowedFilter.php:84]: [Optional]
```suggestion
            $allowedUris[] = TaoOntology::LANGUAGE_PREFIX . $locale;
```

File: models/classes/Language/Filter/LanguageAllowedFilter.php [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        }
+
+        foreach ($collection->getUris() as $uri) {
+            if (!in_array($uri, $allowedUris)) {
```
Comment (shpran) [inline models/classes/Language/Filter/LanguageAllowedFilter.php:88]:
```suggestion
            if (!in_array($uri, $allowedUris, true)) {
```

File: models/classes/Lists/Business/Domain/ValueCollection.php [hunk 6]
```diff
@@ -53,6 +53,21 @@ public function getUri(): ?string
...
 
+    public function removeValueByUri(string $uri): self
+    {
+        /**
+         * @var int $key
+         * @var Value $value
+         */
+        foreach ($this->values as $key => $value) {
+            if ($value->getUri() === $uri) {
+                unset($this->values[$key]);
```
Comment (shpran) [inline models/classes/Lists/Business/Domain/ValueCollection.php:64]: [Optional] Wdyt, maybe after unset we need to reset array indexes?

**Review by shpran — APPROVED** [review]
Good job!
Left some comments

File: migrations/Version202411141300522230_tao.php (hunk 1 above)
Comment (gabrielfs7) [reply migrations/Version202411141300522230_tao.php:23]: I agree, I will do that if no other colleague sees issues

File: models/classes/Language/Filter/LanguageAllowedFilter.php (hunk 4 above)
Comment (gabrielfs7) [reply models/classes/Language/Filter/LanguageAllowedFilter.php:84]: I still prefer the typesafe concatenation in this case, if you do not mind :)

File: models/classes/Lists/Business/Domain/ValueCollection.php (hunk 6 above)
Comment (gabrielfs7) [reply models/classes/Lists/Business/Domain/ValueCollection.php:64]: I cannot, cause the index is an uri

File: models/classes/Language/Filter/LanguageAllowedFilter.php (hunk 3 above)
Comment (gabrielfs7) [reply models/classes/Language/Filter/LanguageAllowedFilter.php]: Done, thanks!

File: migrations/Version202411141300522230_tao.php (hunk 1 above)
Comment (gabrielfs7) [reply migrations/Version202411141300522230_tao.php]: Now it is part of taoUpdate.php

**Pull request merged**